### ISSUE TITLE
Array.pick_random() is not uniform (very minor issue)

### ISSUE THREAD

**Issue by HenryLoenwind** [godotengine/godot#82117]
### Godot version

v4.1.1.stable.official [bd6af8e0e]

### System information

Godot v4.1.1.stable - Windows 10.0.19045 - Vulkan (Forward+) - dedicated NVIDIA GeForce RTX 3060 Ti (NVIDIA; 31.0.15.3623) - 11th Gen Intel(R) Core(TM) i5-11400F @ 2.60GHz (12 Threads)

### Issue description

Array.pick_random() uses the same modulo-solution that is suggested in the documentation. However, for array lengths that don't divide the random number's range evenly, this method leads to a distribution that is not uniform but biased towards the lower side of the range.

The reason is simple: When using modulo, the last repetition is cut short. For example, `(0..8) % 5` will produce `0,1,2,3,4, 0,1,2`, thus producing one extra set of `0,1,2`.

With pick_random() (and `a[randi() % a.size()]`) the bias usually is very small, as the random number range typically is much bigger than result range. However, in my opinion, an engine should not be producing biased results (unless there's a clear warning in the documentation), even if it won't affect 99.9% of its users.

Proposed solution: As the bias is very small when using pick_random() due to array sizes being limited by RAM and nowhere near the RNG range, I suggest adding a warning to the documentation about the bias. I find it hard to construct an example where the bias would be noticeable in a game.

### Steps to reproduce

This needs a selection from a huge array and/or millions of calls to be visible. To show that the general effect does exist, using a RNG with a smaller range can be used:

```
	var a = [0,1,2,3,4,5,6,7,8]
	var d = [0,0,0,0,0,0,0,0,0]
	for i in range(100_000_000):
		d[ a[ (randi() & 0xf) % a.size() ] ] += 1
	for i in a:
		print("Number %d got %6.3f%%, %d votes" % [i, d[i] * 100.0 / 100_000_000, d[i]])
```

```
Number 0 got 12.497%, 12497089 votes
Number 1 got 12.501%, 12500656 votes
Number 2 got 12.502%, 12501931 votes
Number 3 got 12.499%, 12498787 votes
Number 4 got 12.498%, 12498340 votes
Number 5 got 12.499%, 12498876 votes
Number 6 got 12.502%, 12502284 votes
Number 7 got  6.252%, 6251961 votes
Number 8 got  6.250%, 6250076 votes
```

As the difference between the RNG range (0..15) and the used range (0..8) is minimal, we can see the bias clearly. The bigger the size differences between the ranges get, the smaller the bias becomes.

### Minimal reproduction project

see above

**Comment by AThousandShips** [comment]
Your example is extremely unfair though, can you show that there's any real disparity in the random results?

**Comment by AThousandShips** [comment]
With some testing with random generation the problem is infinitesimal for any array that is reasonably small, the great disparity only appears if the range has been clamped to very close to the size of your array, as long as the array is at least several orders of magnitude smaller than the range the difference is minimal (on the order of ~0.001%)

For example with a clamp range as in your example that is 16 times larger (4096 compared with an array of 248 elements) the difference is ~0.02%

For that to affect any code you'd have to have an array with 268'435'455 elements

In summary: The errors from this method are likely smaller than the natural variation in the distribution for any practical application

**Comment by lawnjelly** [comment]
More discussion on this here:
https://twitter.com/keenanisalive/status/1702709103494103326

**Comment by AThousandShips** [comment]
The question is: is the extra cost of `random(0, n)` worth it compared to the improvement of distribution (which is miniscule)

**Comment by HenryLoenwind** [comment]
> The question is: is the extra cost of `random(0, n - 1)` worth it compared to the improvement of distribution (which is miniscule)

If this was a general-purpose language, yes. But with it being targeted at games, a sentence of warning is enough, in my opinion. If someone manages to write a game that relies on a uniform distribution, they should also be able to manage that themselves. They just need to know to do that.

**Comment by Mickeon** [comment]
I thought of including this in https://github.com/godotengine/godot/pull/69451, but then...
I'm sorry, the margin of bias towards certain numbers is so miniscule, it really does not matter. At best, it could be added as a pretty small note in the **RandomNumberGenerator** class.
Bloating users with information they may never, ever need to know for general use is something that I personally actively avoid. It makes things look more complicated than they are in practice.

**Comment by EAinsley** [comment]
Even if the array is large enough, the bias should still be minuscule theoretically. The upper bound of the **absolute** bias is at most $`2^{-31}`$, which is about 5e-10 (Suppose the maximum range of RNG is $`2^{32}`$). 
The upper bound of the **relative** bias is $`2`$. However, it's very unlikely to getting any close to the upper bound in practice. As a quick reference, say we tolerate a relative bias of $`10^{-3}`$, it is safe to sample from an array of size $`2^{20}`$, which is 1'048'576. 
So if you're not using it to do massive lottery or building cryptography protocols, you are fine. This is actually a pretty good guarantee in most cases.
Although I don't think we need to warn the programmers to pay extra attention, it may be worthwhile to warn people not to use this in a security critical situation.

---

## Proofs
Say that we want to sample from an array of size $`n`$. Consider the probability of getting the $`k`$-th element in the array, $`k \in [0, n)`$. Also, denote the number given by the random generator $`c`$, where $`c \in [0, M)`$. 
Let $`C = \{c \in [0, M) \mid c= k + zn, z \in \mathbb Z\}`$. Getting $`k`$ eventually is equivalent to $`c \in C`$. The probability of such event is 
$$p = \frac{|C|}{M}$$
where
$$|C| = \lfloor\frac{M -k}{n}\rfloor$$
Now we can calculate the absolute bias and relative bias. 

### Absolute
The expected probability of getting an element is $\frac1n$. To measure the maximum bias is to maximize $`|p - \frac1n|`$

$$\begin{aligned}
\max|p - \frac1n| &= \max |\lfloor\frac{M -k}{n}\rfloor\cdot\frac1M - \frac1n| \\
& \leq\left(\max |\frac{M -k}{n}\cdot\frac1M - \frac{1}{n} | + \frac1M\right) \\
& =\left(\max |\frac{k}{nM} | + \frac1M\right)\\
& =\left(\max |\frac{n - 1}{n} | + 1\right)\cdot\frac1M \\
& =\left(\max |1 - \frac1n | + 1\right)\cdot\frac1M \\
& \leq \frac2M \\
\end{aligned}$$

In this case, we are using 32-bit unsigned integer, so $`M = 2^{32}`$ and the result is $`2^{-31}`$.

### Relative
The expected probability of getting any element is $`\frac1n`$, so the relative bias is $`(p - \frac1n)/\frac1n`$. The result is 
$$\frac {2n}{M}$$. The maximum value is therefore $`2`$ when $`n`$ is $`2^{32}`$.
This is significant theoretically, it's very unlikely to encounter such extremity in practice. If you can tolerant at most $`b`$ relative bias, then the maximum array size it can support is
$$\frac {bM}2$$

**Comment by AThousandShips** [comment]
Thank you for adding some mathematical context! I'd say people shouldn't use any of these functions for sensitive data generation and should use the cryptographic stuff for anything of that kind

**Comment by AThousandShips** [comment]
It'd be interesting to know what the actual theoretical bias of the RNG itself, the code for it is here:
https://github.com/godotengine/godot/blob/6649a84f6225bee93990752bc0aaf8f512f196d1/thirdparty/misc/pcg.cpp#L6-L15

**Comment by EAinsley** [comment]
It would be way too difficult to do theoretical analysis of the RNG, but we can find empirical analysis in [the original paper](https://www.pcg-random.org/pdf/toms-oneill-pcg-family-v1.02.pdf). Please see section 2.1 for statistical qualities and section 6 for performance. In section 6.3.1 (and also in the code comments), we can find the used implementations should be PCG-XSH-RR (32-bit output with 64-bit state). The potential non-uniformity of this specific PCG is undetectable under BigCrush from TestU01 (Figure 15).
